### PR TITLE
Fix 'attempting to reference a deleted function' with MSVC

### LINKED ISSUE #8192: MSVC Error, attempting to reference a deleted function
I am trying to rebuild chainerx with latest everything (os, compiler, cuda/cudnn),
I also modified the `setup.py` trying to build with ninja generator. And encounter the following error:
```
C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1423~1.281\bin\Hostx64\x64\cl.exe  /nologo /TP -DCHAINERX_ENABLE_CUDA=1 -DCHAINERX_ENABLE_LAPACK=1 -DWIN32_LEAN_AND_MEAN -Dgsl_api="" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -Ipybind11\include -Igsl-lite\include -Iabseil -Igoogletest-src\googletest\include -ID:\chainer\chainerx_cc -DVERSION_INFO=\"7.0.0b4\"  /DWIN32 /D_WINDOWS /W3 /GR /EHsc /openmp /MD /O2 /Ob2 /DNDEBUG   -std:c++14 /showIncludes /Fochainerx\routines\CMakeFiles\chainerx_routines.dir\connection.obj /Fdchainerx\routines\CMakeFiles\chainerx_routines.dir\chainerx_routines.pdb /FS -c D:\chainer\chainerx_cc\chainerx\routines\connection.cc
D:\chainer\chainerx_cc\chainerx\routines\connection.cc(64): warning C4244: 'argument': conversion from 'int64_t' to 'int8_t', possible loss of data
D:\chainer\chainerx_cc\chainerx\routines\connection.cc(365): error C2280: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': attempting to reference a deleted function
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: see declaration of 'chainerx::BackwardBuilder::BackwardBuilder'
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': function was explicitly deleted
```

https://github.com/chainer/chainer/blob/f653368d7f296878360f3d98ff1c128fb2a51e1d/chainerx_cc/chainerx/routines/connection.cc#L365-L366
where 
https://github.com/chainer/chainer/blob/f653368d7f296878360f3d98ff1c128fb2a51e1d/chainerx_cc/chainerx/backward_builder.h#L141

The error seems to be valid.

### PR DESCRIPTION
**Pull request by cloudhan**
Fixed #8192, Close #8193 

This PR will suppress #8193.

noexcept(true) is not required for std::unordered_map, causing compilation error with MSVC for noexcept moveable BackwardBuilder. For more detail, please refers to https://github.com/microsoft/STL/issues/165

So a simple fix is just replace `std::unordered_map` with noexcept guaranteed hash map

Note, according to https://www.ece.uvic.ca/~frodo/cppdraft/n4659/html/dcl.fct.def#delete-2
> A program that refers to a deleted function implicitly or explicitly, other than to declare it, is ill-formed. [ Note: This includes calling the function implicitly or explicitly and forming a pointer or pointer-to-member to the function. It applies even for references in expressions that are not potentially-evaluated. If a function is overloaded, it is referenced only if the function is selected by overload resolution. The implicit odr-use ([basic.def.odr]) of a virtual function does not, by itself, constitute a reference. — end note ]

Deleted move assignment operator is still used, causing the program ill-formed, but seems both GCC and MSVC accept it.

### CODE REVIEW

**Comment by cloudhan** [comment]
@niboshi remove `noexcept` might also do the trick, might need additional minor changes on the call site of move. It's OK to swap std::unordered_map for absl::flat_hash_map for `noexcept`-ness?

**Comment by emcastillo** [comment]
Just for [reference](https://github.com/abseil/abseil-cpp/issues/223)  "Iteration on absl's tables is O(capacity), not O(size)."

Can we measure how this change could impact performance?

**Comment by niboshi** [comment]
Another way around could be to `try { } catch (...) { }` ?

**Comment by cloudhan** [comment]
> Another way around could be to `try { } catch (...) { }` ?

Only if you want to implement move ctor manually.

**Comment by cloudhan** [comment]
FYI, there are interesting sets of benchmarks about hashmap https://martin.ankerl.com/2019/04/01/hashmap-benchmarks-01-overview/

**Comment by emcastillo** [comment]
I will try to measure the effect on chainerX, from the links you posted, it actually seems that we could get more performance!!

**Comment by cloudhan** [comment]
Actually only
https://github.com/chainer/chainer/blob/4c79f50db50fdf165f6f5efa77df57dc4d2bb9d5/chainerx_cc/chainerx/backward_builder.h#L225 is needed, if you don't mind use std and abseil's hash map simultaneously.

**Comment by cloudhan** [comment]
@emcastillo @niboshi Any progress or suggestion?

**Comment by emcastillo** [comment]
I just tested it with imagenet and resnet-50, performance remains exactly the same 👍

File: chainerx_cc/chainerx/backward_builder.h [hunk 1]
```diff
@@ -7,11 +7,11 @@
 #include <memory>
 #include <numeric>
 #include <set>
-#include <unordered_map>
 #include <utility>
 #include <vector>
 
 #include <gsl/gsl>
+#include <absl/container/flat_hash_map.h>
```
Comment (emcastillo) [inline chainerx_cc/chainerx/backward_builder.h]: include order, absl before gsl

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**